packages/instrumentation-llamaindex/src/instrumentation.ts

```diff
@@ -0,0 +1,358 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import type * as llamaindex from "llamaindex";
+
+import { context, trace, SpanStatusCode } from "@opentelemetry/api";
+import {
+  InstrumentationBase,
+  InstrumentationModuleDefinition,
+  InstrumentationNodeModuleDefinition,
+  safeExecuteInTheMiddle,
+} from "@opentelemetry/instrumentation";
+import { SpanAttributes } from "@traceloop/ai-semantic-conventions";
+import { LlamaIndexInstrumentationConfig } from "./types";
+import { BaseEmbedding, BaseSynthesizer, LLM, BaseRetriever } from "llamaindex";
+
+export class LlamaIndexInstrumentation extends InstrumentationBase<any> {
+  protected override _config!: LlamaIndexInstrumentationConfig;
+
+  constructor(config: LlamaIndexInstrumentationConfig = {}) {
+    super("@traceloop/instrumentation-llamaindex", "0.0.17", config);
+  }
+
+  public override setConfig(config: LlamaIndexInstrumentationConfig = {}) {
+    super.setConfig(config);
+  }
+
+  public manuallyInstrument(
+    module: typeof llamaindex & { openLLMetryPatched?: boolean },
+  ) {
+    if (module.openLLMetryPatched) {
+      return;
+    }
+  }
+
+  protected init(): InstrumentationModuleDefinition<any> {
+    const module = new InstrumentationNodeModuleDefinition<any>(
+      "llamaindex",
+      [">=0.0.40"],
+      this.patch.bind(this),
+      this.unpatch.bind(this),
+    );
+    return module;
+  }
+
+  private isLLM(llm: any): llm is LLM {
+    return (
+      llm &&
+      (llm as LLM).complete !== undefined &&
+      (llm as LLM).chat !== undefined
+    );
+  }
+
+  private isEmbedding(embedding: any): embedding is BaseEmbedding {
+    return (
+      embedding instanceof BaseEmbedding &&
+      embedding.getQueryEmbedding !== undefined
+    );
+  }
+
+  private isSynthesizer(synthesizer: any): synthesizer is BaseSynthesizer {
+    return (
+      synthesizer && (synthesizer as BaseSynthesizer).synthesize !== undefined
+    );
+  }
+
+  private isRetriever(retriever: any): retriever is BaseRetriever {
+    return retriever && (retriever as BaseRetriever).retrieve !== undefined;
+  }
+
+  private patch(
+    moduleExports: typeof llamaindex & { openLLMetryPatched?: boolean },
+  ) {
+    if (moduleExports.openLLMetryPatched) {
+      return moduleExports;
+    }
+
+    this._wrap(
+      moduleExports.RetrieverQueryEngine.prototype,
+      "query",
+      this.genericWrapper("RetrieverQueryEngine", "query"),
+    );
+
+    for (const key in moduleExports) {
+      const cls = (moduleExports as any)[key];
+      if (this.isLLM(cls.prototype)) {
+        this._wrap(
+          cls.prototype,
+          "complete",
+          this.completeWrapper({ className: cls.name }),
+        );
+        this._wrap(
+          cls.prototype,
+          "chat",
+          this.chatWrapper({ className: cls.name }),
+        );
+      } else if (this.isEmbedding(cls.prototype)) {
```

Comment:
query embedding, synthesize and retrieve should all start tasks:
https://github.com/traceloop/openllmetry/blob/main/packages/opentelemetry-instrumentation-llamaindex/opentelemetry/instrumentation/llamaindex/base_embedding_instrumentor.py

https://github.com/traceloop/openllmetry/blob/main/packages/opentelemetry-instrumentation-llamaindex/opentelemetry/instrumentation/llamaindex/base_retriever_instrumentor.py

https://github.com/traceloop/openllmetry/blob/main/packages/opentelemetry-instrumentation-llamaindex/opentelemetry/instrumentation/llamaindex/base_synthesizer_instrumentor.py